### ISSUE TITLE
Improve docs regarding crate-type

### ISSUE THREAD

**Issue by lukaslueg** [rust-lang/cargo#4809]
This is a follow-up to #3169. The [docs](http://doc.crates.io/book/reference/manifest.html#building-dynamic-or-static-libraries) currently state

> The available options are `dylib, rlib, staticlib, cdylib`, and `proc-macro`. You should only use this option in a project. *Cargo will always compile packages (dependencies) based on the requirements of the project that includes them.*

The last sentence is actually not quite true in it's absolute: If a dependency specifies only `staticlib` or `cdylib` and the importing crate tries to `extern crate` it, rustc will fail with a confusing error as the dependency provides no linkable target. #4797 added a warning to that tune.

The last sentence should make clear that Cargo is in fact not free to choose whatever is required.

**Comment by durka** [comment]
*Is this a bug in cargo or a doc bug?* Forget "not quite true", the last sentence implies the exact opposite of what cargo does.

If what the doc says is the intention, then cargo should be fixed to ignore the crate-type setting for dependencies. If cargo's actual current behavior is the intention, the italicized sentence should be changed to:

> Be sure to include `rlib` as one of the crate types for a crate intended to be used as a dependency, as otherwise Cargo will get confused when you try to do so.

**Comment by lukaslueg** [comment]
This sentence " Cargo will always compile packages ..." is still in the docs. Does Cargo fall short of what is expected of it or are the docs too flattering? Anyway, bot, be gone!

**Comment by durka** [comment]
Yeah, I am **really** not a fan of this bot, having seen it in other
projects. Issues getting old because there's not enough manpower to fix
them doesn't make the issues invalid.

**Comment by lukaslueg** [comment]
![](https://i.imgur.com/nTJgdsl.jpg)

**Comment by agrif** [comment]
I just ran in to the warning added in #4797, and I want to chime in with my use-case.

Currently, when I depend on a crate that only produces a `cdylib`, running `cargo build` happily builds that dylib and places it in `target/.../deps/`. This is exactly what I want and what I was expecting: I'm trying to write a program that can load and unload plugins, and I would like to be able to pull in and compile those plugins as crates, using cargo. Since these plugin crates are only ever meant to be loaded dynamically, there is no reason for them to provide an `rlib` (and in fact it would probably be bad, since every plugin needs to export a non-mangled symbol of the same name.)

I never try to use `extern crate` on those crates, and so I never run in to the `rustc` issue. I would also be pretty sad if cargo stopped building the `cdylib` because it thought it would never be linked in, and therefore never used. Or if that warning turned in to a hard error, as promised.

This is not a common use case, sure. Maybe it could be solved by adding a flag to dependencies to say "I want all the build products from this, even if it looks like I don't need them."
